src/cli/java/org/commcare/util/screen/EntityScreen.java

```diff
@@ -318,4 +318,18 @@ public String toString() {
     public Hashtable<String, TreeReference> getReferenceMap() {
         return referenceMap;
     }
+
+    @SupressWarnings("unsused")
+    public boolean referencesContainStep(String stepValue) {
+        if (referenceMap != null) {
+            return referenceMap.containsKey(stepValue);
+        }
+        for (TreeReference ref: references) {
+            String id = getReturnValueFromSelection(ref,(EntityDatum) session.getNeededDatum(), evalContext);
```

Comment:
should `session.getNeededDatum()` be pulled out of the loop?

Reply:
Definitely. There's a lot of potential processing in that step.

Reply:
Interesting, I did it this way because that is how it is done when creating the `referenceMap` above. Sounds like we should change that one too?

Reply:
https://github.com/dimagi/commcare-core/blob/formplayer/src/cli/java/org/commcare/util/screen/EntityScreen.java#L113

Reply:
Followup question, is there a reason not to use the `mNeededDatum` attribute in both places, rather than fetch it from the session again.

Reply:
Reusing mNeededDatum would require a bit of a deeper look than I'm familiar with currently.

My strong sense is that this wasn't pulled out in the other loop simply because originally this was toy code for the CLI that wasn't written with performance in mind. I agree that it would be good to pull it out of the loop above as well